src/Cardano/Wallet.hs

```diff
@@ -122,36 +130,48 @@ mkWalletLayer db network = WalletLayer
                 mkAddressPool (publicKey accXPrv) (gap w) ExternalChain []
         let intPool =
                 mkAddressPool (publicKey accXPrv) minBound InternalChain []
-        let wallet = initWallet $ SeqState
+        let wid =
+                WalletId (digest $ publicKey rootXPrv)
+        let checkpoint = initWallet $ SeqState
                 { externalPool = extPool
                 , internalPool = intPool
                 }
-        let wid = WalletId (digest $ publicKey rootXPrv)
-        liftIO (readCheckpoint db (PrimaryKey wid)) >>= \case
-            Nothing -> do
-                liftIO $ putCheckpoint db (PrimaryKey wid) wallet
-                return wid
-            Just _ ->
-                throwE $ ErrCreateWalletIdAlreadyExists wid
-    , readWallet = \wid -> liftIO (readCheckpoint db (PrimaryKey wid)) >>= \case
-        Nothing ->
-            throwE $ ErrReadWalletNotFound wid
-        Just w ->
-            return (w, error "FIXME: store and retrieve wallet metadata")
-
-    , watchWallet = liftIO . listen network . onNextblocks
+        now <- liftIO getCurrentTime
+        let metadata = WalletMetadata
+                { name = Cardano.Wallet.name w
+                , passphraseInfo = WalletPassphraseInfo now
+                , status = Restoring minBound
+                , delegation = NotDelegating
+                }
+        DB.createWallet db (PrimaryKey wid) checkpoint metadata $> wid
+
+    , readWallet = \wid -> maybeToExceptT (ErrNoSuchWallet wid) $ do
+        cp <- MaybeT $ DB.readCheckpoint db (PrimaryKey wid)
+        meta <- MaybeT $ DB.readWalletMeta db (PrimaryKey wid)
+        return (cp, meta)
+
+    , watchWallet =
+        liftIO . listen network . onNextblocks
     }
   where
     onNextblocks :: WalletId -> [Block] -> IO ()
     onNextblocks wid blocks = do
-        (txs, cp') <- readCheckpoint db (PrimaryKey wid) >>= \case
+        (txs, cp') <- DB.readCheckpoint db (PrimaryKey wid) >>= \case
             Nothing ->
                 fail $ "couldn't find worker wallet: " <> show wid
             Just cp -> do
                 let nonEmpty = not . null . transactions
                 return $ applyBlocks (filter nonEmpty blocks) cp
-        putCheckpoint db (PrimaryKey wid) cp'
-        unsafeRunExceptT $ putTxHistory db (PrimaryKey wid) txs -- Safe after ^
+        -- FIXME
+        -- Note that, the two calls below are _safe_ under the assumption that
```

Comment:
Maybe also add a note that there is no transaction surrounding `readCheckpoint` and `putCheckpoint` because there will be only one thread applying blocks.